db/change_cache.go

```diff
@@ -577,31 +577,118 @@ func (c *changeCache) releaseUnusedSequence(ctx context.Context, sequence uint64
 	}
 }
 
-// releaseUnusedSequenceRange calls processEntry for each sequence in the range, but only issues a single notify.
+// releaseUnusedSequenceRange will handle unused sequence range arriving over DCP. It will batch remove from skipped or
+// push a range to pending sequences, or both.
 func (c *changeCache) releaseUnusedSequenceRange(ctx context.Context, fromSequence uint64, toSequence uint64, timeReceived time.Time) {
 
 	base.InfofCtx(ctx, base.KeyCache, "Received #%d-#%d (unused sequence range)", fromSequence, toSequence)
 
 	unusedSeq := channels.NewID(unusedSeqKey, unusedSeqCollectionID)
 	allChangedChannels := channels.SetOfNoValidate(unusedSeq)
-	for sequence := fromSequence; sequence <= toSequence; sequence++ {
+
+	// if range is single value, just run sequence through process entry and return early
+	if fromSequence == toSequence {
 		change := &LogEntry{
-			Sequence:     sequence,
+			Sequence:     toSequence,
 			TimeReceived: timeReceived,
 		}
-
-		// Since processEntry may unblock pending sequences, if there were any changed channels we need
-		// to notify any change listeners that are working changes feeds for these channels
 		changedChannels := c.processEntry(ctx, change)
 		allChangedChannels = allChangedChannels.Update(changedChannels)
 		c.channelCache.AddUnusedSequence(change)
+		if c.notifyChange != nil {
+			c.notifyChange(ctx, allChangedChannels)
+		}
+		return
 	}
 
+	// acquire mutex on cache for reading/writing to pending + reading next sequence value
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if toSequence < c.nextSequence {
+		// batch remove from skipped
+		err := c.RemoveSkippedSequences(ctx, fromSequence, toSequence)
+		if err != nil {
+			base.DebugfCtx(ctx, base.KeyCache, err.Error())
+			// we have sequences (possibly duplicate sequences) in the unused range that don't exist in skipped list
+			seqRange := &channels.UnusedSequenceRange{
+				StartSeq: fromSequence,
+				EndSeq:   toSequence,
+			}
+			// handle any potential duplicate sequences between the unused range and the skipped list
+			err = c.skippedSeqs._removeSubsetOfRangeFromSkipped(ctx, seqRange)
+			if err != nil {
+				base.DebugfCtx(ctx, base.KeyCache, "error processing subset of unused sequences in skipped list: %v", err)
+			}
+		}
+	} else if fromSequence >= c.nextSequence {
+		// whole range to pending
+		seqRange := &channels.UnusedSequenceRange{
+			StartSeq: fromSequence,
+			EndSeq:   toSequence,
+		}
+		c._pushRangeToPending(ctx, seqRange, timeReceived)
+		// unblock any pending sequences we can after new range(s) have been pushed to pending
+		changedChannels := c._addPendingLogs(ctx)
+		allChangedChannels = allChangedChannels.Update(changedChannels)
+		c.internalStats.pendingSeqLen = len(c.pendingLogs)
+	} else {
```

Comment:
I think we want some context here to describe why this indicates a duplicate sequence.
```suggestion
	} else {
	   // An unused sequence range than includes c.nextSequence in the middle of the range
	   // isn't possible under normal processing - unused sequence ranges will normally be moved
	   // from pending to skipped in their entirety, as it's the processing of the pending sequence 
	   // *after* the range that triggers the range to be skipped.  A partial range in skipped means 
	   // an duplicate entry with a sequence within the bounds of the range was previously present 
	   // in pending.  
```